packages/origin-247-certificate/src/offchain-certificate/offchain-certificate.service.ts

```diff
@@ -77,40 +94,55 @@ export class OffchainCertificateService<T = null> {
     }
 
     public async batchClaim(commands: IClaimCommand[]): Promise<void> {
+        const grouped = this.groupCommandsByCertificate(commands);
+
+        for (const certificateId in grouped) {
+            const events = grouped[certificateId].map((c) => {
+                return new CertificateClaimedEvent(parseInt(certificateId), c);
+            });
+            try {
+                this.createAggregate(events);
+            } catch (err) {
+                throw new CertificateErrors.BatchError(err);
+            }
+        }
+
         for (const command of commands) {
             await this.claim(command);
         }
     }
 
     public async batchTransfer(commands: ITransferCommand[]): Promise<void> {
+        const grouped = this.groupCommandsByCertificate(commands);
+
+        for (const certificateId in grouped) {
+            const events = grouped[certificateId].map((c) => {
+                return new CertificateTransferredEvent(parseInt(certificateId), c);
+            });
+            try {
+                this.createAggregate(events);
+            } catch (err) {
+                throw new CertificateErrors.BatchError(err);
+            }
+        }
+
         for (const command of commands) {
             await this.transfer(command);
         }
     }
 
-    private async createAggregate(event: ICertificateEvent): Promise<CertificateAggregate> {
-        const readModel = await this.readModelRepo.getByInternalCertificateId(
-            event.internalCertificateId
-        );
-
-        let aggregate: CertificateAggregate;
-
-        if (readModel) {
-            aggregate = CertificateAggregate.fromReadModel(readModel);
-            aggregate.apply(event);
-        } else {
-            aggregate = CertificateAggregate.fromEvents([
-                ...(await this.certEventRepo.getByInternalCertificateId(
-                    event.internalCertificateId
-                )),
-                event
-            ]);
-        }
+    private async createAggregate(events: ICertificateEvent[]): Promise<CertificateAggregate> {
+        const aggregate = CertificateAggregate.fromEvents([
+            ...(await this.certEventRepo.getByInternalCertificateId(
+                events[0].internalCertificateId
+            )),
+            ...events
+        ]);
         return aggregate;
     }
 
     private generateInternalCertificateId(): number {
-        return new Date().getTime();
+        return Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
```

Comment:
What in case of conflict (already existing entry)? We have to handle that unfortunately. Can't we just have increasing number or whatever?

Reply:
I've changed the implementation to take the number of issuance events and return that increased by one.